### ISSUE TITLE
BUG: Create app command missing `.gitignore` file

### ISSUE THREAD

**Issue by nicodes** [QwikDev/astro#166]
When creating a project with the `pnpm create @qwikdev/astro` command there is no `.gitignore` file created. When selecting the `Initialize a new git repository?` option in the creation flow this causes some directories to be committed that shouldn't be (like `node_modules`).

**Comment by siguici** [comment]
Indeed @nicodes, I just realized that some template files are not copied into the project directory.
This is good because I'm currently working on improving the CLI so I'll take the opportunity to fix this issue.

**Comment by thejackshelton** [comment]
@siguici did we ever add this in? If not I can go ahead

**Comment by siguici** [comment]
> @siguici did we ever add this in? If not I can go ahead

Not yet. [I'm still working on the CLI](https://github.com/siguici/qwik-astro). I'm trying to refactor the code so that I can add unit tests. There's also issue #167 that I plan to fix. If you can provide fixes while I'm finishing, I'll merge your changes.
It would also be good to update dependencies, if there are no breaking changes.

**Comment by thejackshelton** [comment]
> > @siguici did we ever add this in? If not I can go ahead
> 
> Not yet. [I'm still working on the CLI](https://github.com/siguici/qwik-astro). I'm trying to refactor the code so that I can add unit tests. There's also issue #167 that I plan to fix. If you can provide fixes while I'm finishing, I'll merge your changes. It would also be good to update dependencies, if there are no breaking changes.

I have a bit of time off this week so I will try and get our new changes in with the latest dependencies 😄

**Comment by siguici** [comment]
> I have a bit of time off this week so I will try and get our new changes in with the latest dependencies 😄

Great 😊. I'm pretty busy these days, that's why it's dragging on a bit on my end. 😪

**Comment by siguici** [comment]
I'll have some time tomorrow, I'll take care of it.

**Comment by thejackshelton** [comment]
Awesome.

```
npm i https://pkg.pr.new/QwikDev/astro/@qwikdev/create-astro@cc0fb27
```

bit of a side note:

how can we test the cli with pkg pr new?

I think it's ready to try out with the rewrite, but I can't seem to get this new version to install

**Comment by siguici** [comment]
The CLI should be testable using [its API](https://github.com/QwikDev/astro/blob/main/libs/create-qwikdev-astro/README.md#-api). However, I'm simplifying its usage to make it easier to test and maintain.

**Comment by thejackshelton-kunaico** [comment]
Ok I think the integration is 100% ready to release.

Is it not possible to test a certain CLI version directly rather than via a test?

**Comment by thejackshelton** [comment]
Ok I think the integration is 100% ready to release. Let me know if you need some help 😄 

side note: Is it not possible to test a certain CLI version directly rather than via a test?

**Comment by thejackshelton** [comment]
Also where can I see the current progress on the CLI? Maybe I might be able to unblock us somewhere

**Comment by siguici** [comment]
> side note: Is it not possible to test a certain CLI version directly rather than via a test?

We can test the CLI using subprocesses (spawn).

> Also where can I see the current progress on the CLI? Maybe I might be able to unblock us somewhere

I'm working on the `main` branch [here](https://github.com/siguici/qwik-astro).

I would like you to be able to merge what you have already done or tell me what branch you are working on so I can merge your updates.

**Comment by thejackshelton** [comment]
Just merged into main 😄

**Comment by siguici** [comment]
> Just merged into main 😄

I've merged the updates and am currently making some adjustments to proceed with CLI testing.

**Comment by thejackshelton** [comment]
Awesome. Let me know when you think it's ready to test! Happy to help if ya need it.

I just went ahead and merged in the new buffering system, so I think we are ready for 0.7 once the CLI is in a good place. This is a big update.

Pairing with Luiz on a more deep testing integration with Astro's lifecycle tomorrow morning for the main integration tests. (e.g the build output, Astro + Node, Astro + Deno, Astro + React, etc.)

**Comment by siguici** [comment]
I’ve been trying to find a reliable solution to test the CLI in interactive mode, but so far, all attempts have been in vain. I’ve experimented with various approaches, transitioning from `node:child_process` to `cross-spawn`, then to `execa`, and now even using `node-pty`. Unfortunately, the problem persists: the terminal remains stuck after writing inputs, seemingly waiting for additional entries, which also affects the parent terminal.

Given these challenges, I’m considering shifting focus and testing only the final outcomes of the program instead. This would include verifying:

- Whether the CLI executes successfully and creates all the necessary directories and files for the templates.
- Whether the commands for installing dependencies and initializing Git run without issues.
- ... (maybe other things in addition.)

**Comment by thejackshelton** [comment]
Yep, as long as it outputs the correct things I think we're good. So just so I understand, with package pr new there's no way to test the cli version, and we need to check that the output is correct before a new release?

**Comment by siguici** [comment]
Indeed, this is what I'm currently working on. I'm trying to make the CLI as testable as possible both on the API side and on the program execution side.

**Comment by thejackshelton** [comment]
What's our gameplan for a new release? 

Side note: At some point we'll need to change the package names to `@qwik.dev/astro` etc.

**Comment by siguici** [comment]
> What's our gameplan for a new release?

I plan to complete testing the CLI build this weekend. If the user doesn't specify a destination directory or if the command is run in an existing directory, they will now be prompted to add `@qwikdev/astro` to their project using the `astro add @qwikdev/astro` command.

Feature-wise, everything is ready, and all bugs have been fixed. The only remaining task is to add the tests, which are currently causing the workflow to fail.

> Side note: At some point we'll need to change the package names to `@qwik.dev/astro` etc.

What’s the reason for switching from `@qwikdev` to `@qwik.dev`?

**Comment by siguici** [comment]
When testing CLI execution, would you recommend focusing on a CI workflow approach or unit tests? Which do you think is more effective for ensuring reliability and functionality?

**Comment by thejackshelton** [comment]
> > What's our gameplan for a new release?
> 
> I plan to complete testing the CLI build this weekend. If the user doesn't specify a destination directory or if the command is run in an existing directory, they will now be prompted to add `@qwikdev/astro` to their project using the `astro add @qwikdev/astro` command.
> 
> Feature-wise, everything is ready, and all bugs have been fixed. The only remaining task is to add the tests, which are currently causing the workflow to fail.

Awesome! Does that mean we can now use the CLI even for existing projects? 👀 

> What’s the reason for switching from `@qwikdev` to `@qwik.dev`?

The Qwik domain name is qwik.dev, so mostly for SEO purposes / conciseness I think. This was a decision made by the core team, where as we were the first package in the organization to use the scope. 

I'm all for it, because when we get the docs up the `qwik.dev/astro` can redirect to wherever we want. Or we could have the docs there, so you can know the location of the docs based on the module.

**Comment by thejackshelton** [comment]
> When testing CLI execution, would you recommend focusing on a CI workflow approach or unit tests? Which do you think is more effective for ensuring reliability and functionality?

Probably unit tests, because then we can keep it more modular, fast test times, and easier to find where the issues are. I think we already have vitest in the monorepo.

**Comment by siguici** [comment]
> Awesome! Does that mean we can now use the CLI even for existing Astro projects? 👀

Indeed, the command now supports `astro add @qwikdev/astro`, and the advantage is that we can resume the process (if everything went smoothly) and continue execution. The only issue is that users must have installed the dependencies beforehand, particularly Astro; otherwise, the installation will fail.

Users can directly add it (without having to create a new project) using the `--add` or `-a` option, such as:  
`pnpm create @qwikdev/astro --add`.  
This will skip all steps up to adding the CI and initializing Git.


> A pet peeve of mine has always been the Astro CLI does not handle tsconfig configuration, and I tried to contribute there but it's very messy and didn't seem worth it just to change the jsxImportSource. If we can add our tsconfig setup that'd be amazing.

Would you like us to enable modifications to the `tsconfig`? I could take inspiration from how Astro makes its changes.

I finally tested the API functionality, and everything works as expected. Now, I am moving on to testing the execution.

Currently, the CLI is ready for testing with [**pkg.pr.new**](https://pkg.pr.new/template/0218fc07-ed38-4670-9810-8d70d8e7bb0f). Just import it and pass the arguments. Example:

```shell
pnpm i https://pkg.pr.new/QwikDev/astro/@qwikdev/create-astro@178
```

```typescript
import createQwikDevAstro from "@qwikdev/create-astro";

createQwikDevAstro(args);
```

**Comment by siguici** [comment]
> The Qwik domain name is qwik.dev, so mostly for SEO purposes / conciseness I think. This was a decision made by the core team, where as we were the first package in the organization to use the scope.

This is a very strategic decision, and I deeply appreciate it. It benefits both us and the users: simply searching for qwik.dev/astro leads directly to the package and its documentation. Perfect!

Does this mean that the core of Qwik will now be under the name qwik.dev or qwik.dev/core?


> Probably unit tests, because then we can keep it more modular, fast test times, and easier to find where the issues are. I think we already have vitest in the monorepo.


Indeed, unit testing makes debugging much easier. I plan to adopt both approaches:

- Unit tests to ensure the CLI works as expected.
- Workflow CI to verify functionality across all environments (Windows, Mac, Linux), runtimes (Node.js, Deno, Bun), and package managers (NPM, PNPM, Yarn).

**Comment by siguici** [comment]
I think it would also be a good idea to allow users to start with an Astro template by running `create astro`, then `astro add` from our CLI. Finally, we could make our modifications to perfect the setup.

**Comment by siguici** [comment]
Everything is ready to be tested now.

**Comment by thejackshelton** [comment]
> > The Qwik domain name is qwik.dev, so mostly for SEO purposes / conciseness I think. This was a decision made by the core team, where as we were the first package in the organization to use the scope.
> 
> This is a very strategic decision, and I deeply appreciate it. It benefits both us and the users: simply searching for qwik.dev/astro leads directly to the package and its documentation. Perfect!
> 
> Does this mean that the core of Qwik will now be under the name qwik.dev or qwik.dev/core?
> 
> > Probably unit tests, because then we can keep it more modular, fast test times, and easier to find where the issues are. I think we already have vitest in the monorepo.
> 
> Indeed, unit testing makes debugging much easier. I plan to adopt both approaches:
> 
> * Unit tests to ensure the CLI works as expected.
> * Workflow CI to verify functionality across all environments (Windows, Mac, Linux), runtimes (Node.js, Deno, Bun), and package managers (NPM, PNPM, Yarn).

Yep, the core will be qwik.dev/core -> which will point to the core documentation.

**Comment by thejackshelton** [comment]
> Everything is ready to be tested now.

What would be the best way to test the cli creation on the user side? pkg pr new?

**Comment by thejackshelton** [comment]
Tried the following from a recent pkg pr new commit from 4 hours ago:

jackshelton@MacBook-Pro my-test % pnpm create @qwikdev/astro@https://pkg.pr.new/QwikDev/astro/@qwikdev/create-astro@ad9c184
┌  Let's create a  QwikDev/astro App  ✨
│
◇  Creating new project in  /Users/jackshelton/dev/playground/my-test  ... 🐇
An error occurred during QwikDev/astro project creation: TypeError: Cannot read properties of undefined (reading 'trim')
    at J (file:///Users/jackshelton/Library/Caches/pnpm/dlx/wyzsbf7ymeoch53cxjcts5jj7u/193f9d8e941-1303a/node_modules/.pnpm/@qwikdev+create-astro@https+++pkg.pr.new+QwikDev+astro+@qwikdev+create-astro@ad9c184_sx4akhmlsmv5g4ni7s4l7fuhvq/node_modules/@qwikdev/create-astro/dist/chunk-VDJILD5V.js:1:1337)
    at z (file:///Users/jackshelton/Library/Caches/pnpm/dlx/wyzsbf7ymeoch53cxjcts5jj7u/193f9d8e941-1303a/node_modules/.pnpm/@qwikdev+create-astro@https+++pkg.pr.new+QwikDev+astro+@qwikdev+create-astro@ad9c184_sx4akhmlsmv5g4ni7s4l7fuhvq/node_modules/@qwikdev/create-astro/dist/chunk-VDJILD5V.js:1:1699)
    at A (file:///Users/jackshelton/Library/Caches/pnpm/dlx/wyzsbf7ymeoch53cxjcts5jj7u/193f9d8e941-1303a/node_modules/.pnpm/@qwikdev+create-astro@https+++pkg.pr.new+QwikDev+astro+@qwikdev+create-astro@ad9c184_sx4akhmlsmv5g4ni7s4l7fuhvq/node_modules/@qwikdev/create-astro/dist/chunk-VDJILD5V.js:1:1887)
    at k.updatePackageJson (file:///Users/jackshelton/Library/Caches/pnpm/dlx/wyzsbf7ymeoch53cxjcts5jj7u/193f9d8e941-1303a/node_modules/.pnpm/@qwikdev+create-astro@https+++pkg.pr.new+QwikDev+astro+@qwikdev+create-astro@ad9c184_sx4akhmlsmv5g4ni7s4l7fuhvq/node_modules/@qwikdev/create-astro/dist/chunk-E3OIVSM3.js:2:130)
    at k.execute (file:///Users/jackshelton/Library/Caches/pnpm/dlx/wyzsbf7ymeoch53cxjcts5jj7u/193f9d8e941-1303a/node_modules/.pnpm/@qwikdev+create-astro@https+++pkg.pr.new+QwikDev+astro+@qwikdev+create-astro@ad9c184_sx4akhmlsmv5g4ni7s4l7fuhvq/node_modules/@qwikdev/create-astro/dist/chunk-E3OIVSM3.js:1:6466)
    at async k.run (file:///Users/jackshelton/Library/Caches/pnpm/dlx/wyzsbf7ymeoch53cxjcts5jj7u/193f9d8e941-1303a/node_modules/.pnpm/@qwikdev+create-astro@https+++pkg.pr.new+QwikDev+astro+@qwikdev+create-astro@ad9c184_sx4akhmlsmv5g4ni7s4l7fuhvq/node_modules/@qwikdev/create-astro/dist/chunk-QBYWE57O.js:1:2420)

**Comment by siguici** [comment]
> Yep, the core will be qwik.dev/core -> which will point to the core documentation.

Great. Good decision.

> 
> What would be the best way to test the cli creation on the user side? pkg pr new?

```bash
pnpm dlx https://pkg.pr.new/QwikDev/astro/@qwikdev/create-astro@4bac5a3
```


> Tried the following from a recent pkg pr new commit from 4 hours ago:
> 
>```shell
> jackshelton@MacBook-Pro my-test % pnpm create @qwikdev/astro@https://pkg.pr.new/QwikDev/astro/@qwikdev/create-astro@ad9c184
> ┌  Let's create a  QwikDev/astro App  ✨
> │
> ◇  Creating new project in  /Users/jackshelton/dev/playground/my-test  ... 🐇
> An error occurred during QwikDev/astro project creation: TypeError: Cannot read properties of undefined (reading 'trim')
>     at J (file:///Users/jackshelton/Library/Caches/pnpm/dlx/wyzsbf7ymeoch53cxjcts5jj7u/193f9d8e941-1303a/node_modules/.pnpm/@qwikdev+create-astro@https+++pkg.pr.new+QwikDev+astro+@qwikdev+create-astro@ad9c184_sx4akhmlsmv5g4ni7s4l7fuhvq/node_modules/@qwikdev/create-astro/dist/chunk-VDJILD5V.js:1:1337)
>     at z (file:///Users/jackshelton/Library/Caches/pnpm/dlx/wyzsbf7ymeoch53cxjcts5jj7u/193f9d8e941-1303a/node_modules/.pnpm/@qwikdev+create-astro@https+++pkg.pr.new+QwikDev+astro+@qwikdev+create-astro@ad9c184_sx4akhmlsmv5g4ni7s4l7fuhvq/node_modules/@qwikdev/create-astro/dist/chunk-VDJILD5V.js:1:1699)
>     at A (file:///Users/jackshelton/Library/Caches/pnpm/dlx/wyzsbf7ymeoch53cxjcts5jj7u/193f9d8e941-1303a/node_modules/.pnpm/@qwikdev+create-astro@https+++pkg.pr.new+QwikDev+astro+@qwikdev+create-astro@ad9c184_sx4akhmlsmv5g4ni7s4l7fuhvq/node_modules/@qwikdev/create-astro/dist/chunk-VDJILD5V.js:1:1887)
>     at k.updatePackageJson (file:///Users/jackshelton/Library/Caches/pnpm/dlx/wyzsbf7ymeoch53cxjcts5jj7u/193f9d8e941-1303a/node_modules/.pnpm/@qwikdev+create-astro@https+++pkg.pr.new+QwikDev+astro+@qwikdev+create-astro@ad9c184_sx4akhmlsmv5g4ni7s4l7fuhvq/node_modules/@qwikdev/create-astro/dist/chunk-E3OIVSM3.js:2:130)
>     at k.execute (file:///Users/jackshelton/Library/Caches/pnpm/dlx/wyzsbf7ymeoch53cxjcts5jj7u/193f9d8e941-1303a/node_modules/.pnpm/@qwikdev+create-astro@https+++pkg.pr.new+QwikDev+astro+@qwikdev+create-astro@ad9c184_sx4akhmlsmv5g4ni7s4l7fuhvq/node_modules/@qwikdev/create-astro/dist/chunk-E3OIVSM3.js:1:6466)
>     at async k.run (file:///Users/jackshelton/Library/Caches/pnpm/dlx/wyzsbf7ymeoch53cxjcts5jj7u/193f9d8e941-1303a/node_modules/.pnpm/@qwikdev+create-astro@https+++pkg.pr.new+QwikDev+astro+@qwikdev+create-astro@ad9c184_sx4akhmlsmv5g4ni7s4l7fuhvq/node_modules/@qwikdev/create-astro/dist/chunk-QBYWE57O.js:1:2420)
> ```

Fixed now

**Comment by siguici** [comment]
It's better to try with:

```shell
npx https://pkg.pr.new/QwikDev/astro/@qwikdev/create-astro@4bac5a3
```

**Comment by siguici** [comment]
> It's better to try with:
> 
> ```shell
> npx https://pkg.pr.new/QwikDev/astro/@qwikdev/create-astro@4bac5a3
> ```

@nicodes, you can also try to see if that solves your issue.

**Comment by thejackshelton** [comment]
Just tested it out on the latest and everything looks great! I think we are good to do a new release. WDYT?

Some feedback from testing:

Is it possible to have "No" as the default to adding a server adapter? 

Another point of confusion for me was that it seemed like you had to add the project name twice. Looking back at it, I see that's for if we want to add it to an existing project. I wonder if we can simplify that to one step somehow.

**Comment by siguici** [comment]
> When running the preview and the user did not add a server adapter, this is going to error. In this case I think we just want to run Astro preview.

Yes. I had never noticed this. I'll remove this part of the `package.json` from the template.

> Is it possible to have "No" as the default to adding a server adapter?

I'll fix this quickly.
 
> Another point of confusion for me was that it seemed like you had to add the project name twice. Looking back at it, I see that's for if we want to add it to an existing project. I wonder if we can simplify that to one step somehow.

I had also made the same observation. I [fixed](https://github.com/QwikDev/astro/pull/178/commits/3970c2665f72e3081ddd8391c39979f3d6d74477) it!

**Comment by siguici** [comment]
> Just tested it out on the latest and everything looks great! I think we are good to do a new release. WDYT?
> 

Yes. Everything should be good now. I also think we can prepare a new release.

**Comment by thejackshelton** [comment]
I ran the latest commit in pkg pr new and did not see a gitignore file. Perhaps this is only with the "no server" stub?

**Comment by siguici** [comment]
I just tested (#190) the execution for each template and all files are present

**Comment by siguici** [comment]
![Capture d'écran 2024-12-25 113521](https://github.com/user-attachments/assets/76e8f5b0-664a-4607-862d-18c3939e43c9)

The problem comes from publishing to NPM which ignores `.gitignore` files from all directories.

**Comment by siguici** [comment]
This should be fixed in the latest version. Let me know if that's not the case! Happy to re-open if so.
